code/go/0chain.net/smartcontract/storagesc/readpool.go

```diff
@@ -62,76 +64,44 @@ func (rp *readPool) save(sscKey, clientID string, balances cstate.StateContextI)
 	return
 }
 
-func (rp *readPool) moveBlobberCharge(sscKey string, sp *stakePool,
-	ap *allocationPool, value state.Balance, balances cstate.StateContextI) (
-	err error) {
-
-	if value == 0 {
-		return // avoid insufficient transfer
+func (rp *readPool) movePartToBlobber(
+	sscKey string,
+	ap *allocationPool,
+	sp *stakePool,
+	value float64,
+	balances cstate.StateContextI,
+) error {
+	blobberCharge := sp.Settings.ServiceCharge * float64(value)
+	if blobberCharge > float64(ap.Balance) {
+		return fmt.Errorf("allocation pool balance %v not enough to cover reward %v",
+			ap.Balance, blobberCharge)
 	}
+	ap.Balance -= state.Balance(blobberCharge)
+	sp.Reward += state.Balance(blobberCharge)
 
-	var (
-		dw       = sp.Settings.DelegateWallet
-		transfer *state.Transfer
-	)
-	transfer, _, err = ap.DrainPool(sscKey, dw, value, nil)
-	if err != nil {
-		return fmt.Errorf("transferring tokens read_pool() -> "+
-			"blobber_charge(%s): %v", dw, err)
-	}
-	if err = balances.AddTransfer(transfer); err != nil {
-		return fmt.Errorf("adding transfer: %v", err)
-	}
+	valueLeft := value - blobberCharge
 
-	// blobber service charge
-	sp.Rewards.Charge += value
-	return
-}
-
-func (rp *readPool) movePartToBlobber(sscKey string, ap *allocationPool,
-	sp *stakePool, value state.Balance, balances cstate.StateContextI) (
-	err error) {
-
-	var blobberCharge state.Balance
-	blobberCharge = state.Balance(sp.Settings.ServiceCharge * float64(value))
-	err = rp.moveBlobberCharge(sscKey, sp, ap, blobberCharge, balances)
-	if err != nil {
-		return
-	}
-
-	value = value - blobberCharge // left for stake holders
-
-	if value == 0 {
-		return // avoid insufficient transfer
+	if valueLeft == 0 {
+		return nil
 	}
 
 	var stake = float64(sp.stake())
-	for _, dp := range sp.orderedPools() {
-		var ratio float64
-		if stake == 0.0 {
-			ratio = float64(dp.Balance) / float64(len(sp.Pools))
-		} else {
-			ratio = float64(dp.Balance) / stake
-		}
+	if stake == 0 {
+		return errors.New("no stake pools found")
+	}
 
-		var (
-			move     = state.Balance(float64(value) * ratio)
-			transfer *state.Transfer
-		)
-		transfer, _, err = ap.DrainPool(sscKey, dp.DelegateID, move, nil)
-		if err != nil {
-			return fmt.Errorf("transferring tokens read_pool() -> "+
-				"stake_pool_holder(%s): %v", dp.DelegateID, err)
+	for _, id := range sp.OrderedPoolIds() {
```

Comment:
Unnecessary ordering?

Reply:
It depends on how stable rounding errors are. I guess we can assume that exactly the same rounding errors occur irrespective of the order in which the loop is enacted?